index.js

```diff
@@ -83,6 +99,133 @@ function attachToBody (options, req, reply, next) {
 }
 
+const allowedTokens = ['[', ']', '{}']
+
+function getSerializedKeyTokens (key) {
+  const tokens = key.split(/(\[|]|{})/).filter(Boolean)
+  return tokens.some(token => allowedTokens.includes(token)) ? tokens : []
+}
+
+function validateSerializedKey (tokens, value) {
+  let isValid = true
+  let token = tokens[0]
+
+  // first token can't be '[', ']', or '{}'
+  if (allowedTokens.includes(token)) isValid = false
+  else if (token === '__proto__' || token === 'constructor') {
+    return `${token} is not allowed as a key name`
+  } else {
+    const bracesEndingCount = tokens.filter(t => t === '{}').length
+
+    // only a single {} special ending is allowed
+    if (bracesEndingCount > 1) isValid = false
+    // {}, if present, must be the last token
+    else if (bracesEndingCount > 0 && tokens[tokens.length - 1] !== '{}') isValid = false
+  }
+
+  const tokensLen = tokens.length
+  let nextToken = tokens[1]
+  let i = 1
+  let hasOpenBracket = false
+
+  while (i < tokensLen) {
+    token = nextToken
+    nextToken = tokens[++i]
+
+    if (token === '__proto__' || token === 'constructor') {
+      return `${token} is not allowed as a key name`
+    }
+
+    // invalid closing bracket
+    if (!hasOpenBracket && token === ']') {
+      isValid = false
+      break
+    }
+    // invalid opening bracket
+    if (hasOpenBracket && token === '[') {
+      isValid = false
+      break
+    }
+
+    hasOpenBracket = token === '[' || (hasOpenBracket && token !== ']')
+    // orphan opening bracket
+    if (hasOpenBracket && nextToken === undefined) isValid = false
+    // {} is not allowed within brackets
+    if (hasOpenBracket && token === '{}') isValid = false
+    // after a closing bracket, only an opening bracket or {} are allowed
+    if (token === ']' && nextToken && !['[', '{}'].includes(nextToken)) {
+      isValid = false
+      break
+    }
+  }
+
+  // when the {} special ending is supplied, the value must be valid JSON
+  if (isValid && tokens[tokens.length - 1] === '{}') {
+    try {
+      secureJSON.parse(value)
+    } catch (e) {
+      return 'the {} special ending was used, but value is not a valid JSON'
+    }
+  }
+
+  return isValid ? null : 'invalid serialized key'
+}
+
+function getSerializedKeySegments (options, body, tokens, value) {
+  const segments = [{ parentRef: body, key: tokens[0] }]
+  const parseArrays = options.enableSpecialNotationKeys !== 'objectsOnly'
+
+  const tokensLen = tokens.length
+  let token = tokens[1]
+  let prevToken = tokens[0]
+  let i = 1
+
+  while (i < tokensLen) {
+    const lastSegment = segments[segments.length - 1]
+    let parentRef
+    let key
+    token = tokens[i]
+    prevToken = tokens[i - 1]
+
+    if (token === ']' && prevToken === '[') {
+      parentRef = lastSegment.parentRef[lastSegment.key] ?? (parseArrays ? [] : {})
+      key = parseArrays ? parentRef.length : Object.keys(parentRef).length
+    } else if (parseArrays && token === ']' && isValidIndex(prevToken)) {
```

Comment:
maybe write two while loops, one for the case parseArrays is true and one for the case, that parseArrays is false?

Reply:
I'm really uncertain whether this kind of change would be worth it. After all, we would sacrifice simplicity, readability and maintainability to favor small performance gains.